<a href="https://colab.research.google.com/github/mweb4100/CS4410-Homework/blob/main/Homework7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
# Homework 7
# Exercises 16.1 and 16.4

import tensorflow as tf
import time

from tensorflow.keras.datasets import mnist, fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

In [6]:
# Exercise 16.1

# Load MNIST
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(60000, 28, 28, 1).astype('float32') / 255
x_test = x_test.reshape(10000, 28, 28, 1).astype('float32') / 255

# MNIST model
mnist_model = Sequential([
    Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(128, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

mnist_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

start = time.time()

mnist_model.fit(
    x_train,
    y_train,
    epochs=5,
    batch_size=64
)

mnist_time = time.time() - start

mnist_loss, mnist_accuracy = mnist_model.evaluate(x_test, y_test)

print("MNIST accuracy:", mnist_accuracy)
print("MNIST training time:", mnist_time, "seconds")

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 123s 129ms/step - accuracy: 0.9595 - loss: 0.1329
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 142s 129ms/step - accuracy: 0.9874 - loss: 0.0412
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 141s 129ms/step - accuracy: 0.9907 - loss: 0.0283
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 142s 129ms/step - accuracy: 0.9937 - loss: 0.0200
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 120s 128ms/step - accuracy: 0.9953 - loss: 0.0148
313/313 ━━━━━━━━━━━━━━━━━━━━ 6s 18ms/step - accuracy: 0.9915 - loss: 0.0270
MNIST accuracy: 0.9915000200271606
MNIST training time: 668.3898162841797 seconds


In [7]:
# Load Fashion-MNIST
(fashion_train, fashion_y_train), (fashion_test, fashion_y_test) = fashion_mnist.load_data()

fashion_train = fashion_train.reshape(60000, 28, 28, 1).astype('float32') / 255
fashion_test = fashion_test.reshape(10000, 28, 28, 1).astype('float32') / 255

# Same model using Fashion-MNIST
fashion_model = Sequential([
    Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(128, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

fashion_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

start = time.time()

fashion_model.fit(
    fashion_train,
    fashion_y_train,
    epochs=5,
    batch_size=64
)

fashion_time = time.time() - start

fashion_loss, fashion_accuracy = fashion_model.evaluate(
    fashion_test,
    fashion_y_test
)

print("Fashion-MNIST accuracy:", fashion_accuracy)
print("Fashion-MNIST training time:", fashion_time, "seconds")

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 125s 132ms/step - accuracy: 0.8345 - loss: 0.4546
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 141s 131ms/step - accuracy: 0.8899 - loss: 0.2994
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 122s 130ms/step - accuracy: 0.9070 - loss: 0.2523
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 142s 130ms/step - accuracy: 0.9196 - loss: 0.2187
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 122s 130ms/step - accuracy: 0.9283 - loss: 0.1918
313/313 ━━━━━━━━━━━━━━━━━━━━ 5s 17ms/step - accuracy: 0.9087 - loss: 0.2530
Fashion-MNIST accuracy: 0.9086999893188477
Fashion-MNIST training time: 652.2661898136139 seconds


Exercise 16.1 Answer: The model performed better on MNIST with 99.15% accuracy compared to 90.87% on Fashion-MNIST. The training times were pretty similar. MNIST took about 668 seconds and Fashion-MNIST took about 652 seconds.

In [8]:
# Exercise 16.4
# Remove the first Dense layer

removed_model = Sequential([
    Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(128, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(10, activation='softmax')
])

removed_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

start = time.time()

removed_model.fit(
    x_train,
    y_train,
    epochs=5,
    batch_size=64
)

removed_time = time.time() - start

removed_loss, removed_accuracy = removed_model.evaluate(x_test, y_test)

print("Accuracy with Dense layer removed:", removed_accuracy)
print("Training time:", removed_time, "seconds")

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 118s 124ms/step - accuracy: 0.9516 - loss: 0.1612
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 117s 125ms/step - accuracy: 0.9844 - loss: 0.0496
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 142s 125ms/step - accuracy: 0.9894 - loss: 0.0344
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 140s 123ms/step - accuracy: 0.9916 - loss: 0.0265
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 142s 124ms/step - accuracy: 0.9934 - loss: 0.0207
313/313 ━━━━━━━━━━━━━━━━━━━━ 7s 22ms/step - accuracy: 0.9906 - loss: 0.0287
Accuracy with Dense layer removed: 0.9905999898910522
Training time: 686.0588276386261 seconds


In [9]:
# Add a Dense layer with 4096 neurons

large_model = Sequential([
    Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(128, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(4096, activation='relu'),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

large_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

start = time.time()

large_model.fit(
    x_train,
    y_train,
    epochs=5,
    batch_size=64
)

large_time = time.time() - start

large_loss, large_accuracy = large_model.evaluate(x_test, y_test)

print("Accuracy with 4096 neuron layer:", large_accuracy)
print("Training time:", large_time, "seconds")

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 403s 428ms/step - accuracy: 0.9659 - loss: 0.1099
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 441s 426ms/step - accuracy: 0.9888 - loss: 0.0365
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 400s 426ms/step - accuracy: 0.9920 - loss: 0.0248
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 442s 426ms/step - accuracy: 0.9941 - loss: 0.0182
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 442s 426ms/step - accuracy: 0.9955 - loss: 0.0148
313/313 ━━━━━━━━━━━━━━━━━━━━ 14s 44ms/step - accuracy: 0.9869 - loss: 0.0457
Accuracy with 4096 neuron layer: 0.9868999719619751
Training time: 2127.038788318634 seconds


Exercise 16.4 Answer: Removing the first Dense layer barely changed the accuracy. It went from 99.15% to 99.06%, and the training time was about the same. Adding the 4096-neuron Dense layer lowered the accuracy to 98.69% and made training much slower. It took about 2,127 seconds compared to 668 seconds for the original model.